File: backend/app/schemas/scraping.py
from pydantic import BaseModel, model_validator

from app.schemas.manual_analysis import ManualAnalysisWorkbook


class ScrapeRequest(BaseModel):
    brand_name: str
    # Country or city the brand operates in, e.g. "Tunisie". Without it a brand-name search
    # matches same-named businesses worldwide and blends them into one sentiment number --
    # a search for "Mytek" returned IT firms in Phoenix and wholesalers in Mexico.
    google_location: str | None = None
    facebook_url: str | None = None
    instagram_url: str | None = None
    # Brand mention keywords to search public Facebook posts & comments by keyword (e.g. "tunisie telecom", "mytek tn")
    keywords: str | None = None
    # Trustpilot company domain, e.g. "orange.fr". Optional: skipped when not provided,
    # because resolving it from the brand name alone is unreliable.
    trustpilot_domain: str | None = None
    # Restrict Trustpilot reviews to a recent window so the scrape matches the reporting period.
    trustpilot_period: str | None = None  # "", last30days, last3months, last6months, last12months
    manual_analysis: ManualAnalysisWorkbook | None = None


class ReviewClassification(BaseModel):
    sentiment: str  # "positive", "neutral", "negative"
    confidence: float
    themes: list[str]
    complaint_category: str | None = None


class ScrapedReview(BaseModel):
    platform: str
    author: str | None = None
    text: str
    rating: float | None = None
    date: str | None = None
    url: str | None = None
    classification: ReviewClassification | None = None


class PlatformResult(BaseModel):
    platform: str
    status: str  # "success", "error", "empty"
    review_count: int = 0
    reviews: list[ScrapedReview] = []
    error_message: str | None = None
    # Optional platform-level headline stat, e.g. "TrustScore 1.6/5 · 12,665 reviews".
    summary_stat: str | None = None
    # Share of scraped reviews the brand publicly replied to (Trustpilot exposes this).
    reply_rate_pct: float | None = None
    # Which real-world places/pages the reviews came from, so a consultant can verify the
    # scrape hit the right company rather than a same-named business elsewhere.
    matched_places: list[str] = []
    # Places dropped because they sat outside the requested location.
    excluded_places: list[str] = []


class TopicCount(BaseModel):
    topic: str
    count: int


class AnalysisSummary(BaseModel):
    positive_pct: float
    neutral_pct: float
    negative_pct: float
    top_themes: list[str]
    top_complaints: list[TopicCount]
    summary_text: str | None = None


class ChannelReportSection(BaseModel):
    channel: str
    summary: str
    strengths: list[str] = []
    friction_points: list[str] = []
    key_stats: list[str] = []  # short "label: value" style stat lines, e.g. "Engagement rate: 3.1%"


class FrictionPoint(BaseModel):
    theme: str
    source: str  # "scraped" | "consultant" | "both"
    description: str
    supporting_evidence: list[str] = []
    # The ranked label this point maps to, echoed back by the LLM from the list it was
    # given. Used to attach real volume in code -- so "billing errors" carries "34 mentions,
    # 28% of negative reviews" instead of an unquantified assertion.
    related_label: str | None = None
    mention_count: int | None = None
    share_pct: float | None = None


class SentimentBreakdown(BaseModel):
    """Deterministic sentiment split computed from the AI-classified scraped reviews
    (not produced by the report LLM, so the numbers are trustworthy)."""
    positive_pct: float
    neutral_pct: float
    negative_pct: float
    classified_count: int


class PlatformSentiment(BaseModel):
    """Per-channel sentiment, so the report can name which channel is worst rather than
    averaging every channel into one meaningless number."""
    platform: str
    positive_pct: float
    neutral_pct: float
    negative_pct: float
    classified_count: int
    review_count: int
    avg_rating: float | None = None
    summary_stat: str | None = None
    reply_rate_pct: float | None = None


class RatingDistribution(BaseModel):
    """Star histogram across all rated reviews. A 1.8 average made of 1s and 5s is a very
    different problem from a 1.8 average made entirely of 2s."""
    one: int = 0
    two: int = 0
    three: int = 0
    four: int = 0
    five: int = 0
    total: int = 0
    average: float | None = None


class ThemeFrequency(BaseModel):
    """A theme or complaint category with the volume behind it."""
    label: str
    count: int
    share_pct: float
    negative_pct: float = 0.0
    # Which channels the mentions came from, e.g. {"Google Maps": 6, "Trustpilot": 2}.
    # A consultant needs to know where a complaint surfaces before they can act on it.
    by_platform: dict[str, int] = {}


class ThemeVerbatim(BaseModel):
    """A real customer quote backing a ranked theme, with its provenance intact."""
    label: str
    platform: str
    text: str
    rating: float | None = None
    date: str | None = None
    sentiment: str | None = None
    # Link to the original review, so a quote can be verified at source rather than trusted.
    url: str | None = None


class RatingSentimentCheck(BaseModel):
    """Where the star rating and what the customer wrote disagree.

    Computed only over reviews that carry both a rating and a classified comment. A high
    rating with a negative comment (or the reverse) is a known signal -- rating inflation,
    sarcasm, or review-gaming -- and the size of the gap is itself a CX finding.
    """

    both_count: int = 0  # reviews with a rating AND a classified comment
    high_rating_negative: int = 0  # rated >= 4 but the text reads negative
    low_rating_positive: int = 0  # rated <= 2 but the text reads positive
    mismatch_count: int = 0
    mismatch_pct: float = 0.0


class TrendPoint(BaseModel):
    """One period on the time axis.

    `negative_pct` and `avg_rating` are null when that period's sample is too small to state
    them -- a break in the line is honest, a plotted point on n=3 is not.
    """

    period: str  # "2026-07" or "2026-Q3"
    label: str  # "Jul 2026" / "Q3 2026"
    reviews: int = 0
    classified_count: int = 0
    negative_pct: float | None = None
    avg_rating: float | None = None


class TrendSeries(BaseModel):
    """Present only when the collected reviews can support a time axis at all."""

    granularity: str  # "month" | "quarter"
    points: list[TrendPoint] = []
    # Which metric cleared the sample floor often enough to be worth plotting.
    has_sentiment: bool = False
    has_rating: bool = False
    min_sample: int = 10


class TopicMovement(BaseModel):
    """One complaint category's change in share between the earlier and recent windows.

    Shares are of all classified feedback in each window, so a rise means the category grew
    as a proportion of what customers talked about -- not merely that more reviews arrived.
    """

    label: str
    earlier_count: int
    recent_count: int
    earlier_share: float  # % of classified reviews in the earlier window
    recent_share: float
    delta: float  # recent_share - earlier_share, in percentage points
    direction: str  # "rising" | "falling" | "stable"


class TopicMomentum(BaseModel):
    """Which complaints are gaining or losing ground. Present only when each half of the
    history carries a real sample -- otherwise a movement would just be noise."""

    earlier_label: str  # e.g. "Aug 2025 – Jan 2026"
    recent_label: str
    earlier_total: int  # classified reviews in the earlier window
    recent_total: int
    items: list[TopicMovement] = []


class ReportEvidence(BaseModel):
    """The quantitative backbone of the report. Every field is computed in Python from the
    classified reviews -- the LLM writes the narrative around these numbers, never them."""
    reviews_collected: int = 0
    # Reviews carrying an actual comment. The rest are star-only ratings, which count toward
    # the distribution but have nothing to classify.
    reviews_with_text: int = 0
    reviews_classified: int = 0
    overall_sentiment: SentimentBreakdown | None = None
    platform_sentiment: list[PlatformSentiment] = []
    rating_distribution: RatingDistribution | None = None
    top_themes: list[ThemeFrequency] = []
    top_complaints: list[ThemeFrequency] = []
    verbatims: list[ThemeVerbatim] = []
    # Null whenever review volume per period is too thin to plot -- the report then simply
    # has no trend section rather than showing sampling noise as a trend.
    trend: TrendSeries | None = None
    # Which complaint categories are gaining or losing share over the collected history.
    # Null when each half of the history is too thin for the comparison to mean anything.
    momentum: TopicMomentum | None = None
    # Reviews-weighted share of reviews the brand publicly replied to, across the channels
    # that expose it (Trustpilot). None when no channel reports it.
    reply_rate_pct: float | None = None
    # How many reviews that rate rests on -- a 0% on 8 reviews must not read as covering 200.
    reply_rate_base: int = 0
    # Disagreement between star rating and comment sentiment; None when too few reviews carry
    # both to say anything.
    rating_sentiment: RatingSentimentCheck | None = None


class _CoercibleFromString(BaseModel):
    """Accepts either the structured object or the bare string it used to be.

    Both `top_strengths` and `recommendations` shipped as `list[str]`. Reports already on
    disk still hold strings, and a model that ignores the schema will send them too, so the
    plain form has to keep validating rather than raising.
    """

    text: str

    @model_validator(mode="before")
    @classmethod
    def _accept_plain_string(cls, value: object) -> object:
        if isinstance(value, str):
            return {"text": value}
        return value


class Strength(_CoercibleFromString):
    """A cross-channel strength, carrying the same evidence a friction point does.

    Left unquantified, a positive sitting beside a friction point badged '47 mentions · 16%'
    reads as opinion next to fact.
    """

    related_label: str | None = None
    mention_count: int | None = None
    share_pct: float | None = None


class Recommendation(_CoercibleFromString):
    """An action, tied to the friction point it resolves.

    A recommendation the reader cannot trace back to a measured problem is the least
    defensible thing in a client deliverable, and it is the section they act on.
    """

    # Theme of the friction point this addresses; matched against the report's own points.
    addresses: str | None = None
    mention_count: int | None = None
    share_pct: float | None = None


class DetailedSocialReport(BaseModel):
    executive_summary: str
    overall_sentiment: SentimentBreakdown | None = None  # populated in code from scraped reviews
    evidence: ReportEvidence | None = None  # populated in code; never generated by the LLM
    channel_breakdown: list[ChannelReportSection] = []
    top_strengths: list[Strength] = []
    top_friction_points: list[FrictionPoint] = []
    website_assessment: str | None = None
    recommendations: list[Recommendation] = []
    methodology_note: str


class ScrapingResponse(BaseModel):
    brand_name: str
    scraped_at: str
    platforms: list[PlatformResult]
    total_reviews: int
    analysis: AnalysisSummary | None = None
    detailed_report: DetailedSocialReport | None = None
    # The consultant's raw workbook is echoed back and saved with the report. Filling it in
    # is an hour of work, and without this it survived only in browser storage: a cleared
    # browser lost it, past reports could not be regenerated against an improved prompt,
    # and there was no record of where a figure in the report came from.
    manual_analysis: ManualAnalysisWorkbook | None = None
    # The saved report's filename (set at archive time). Lets the UI address this report --
    # e.g. to open the report chatbot against it -- for both fresh runs and reloaded archives.
    report_filename: str | None = None


class CompanyOption(BaseModel):
    id: int
    name: str

File: frontend/src/types/scraping.ts
// Mirrors backend/app/schemas/scraping.py

import type { ManualAnalysisWorkbook } from "./manual-analysis";

export interface ReviewClassification {
  sentiment: "positive" | "neutral" | "negative";
  confidence: number;
  themes: string[];
  complaint_category: string | null;
}

export interface ScrapedReview {
  platform: string;
  author: string | null;
  text: string;
  rating: number | null;
  date: string | null;
  url: string | null;
  classification?: ReviewClassification | null;
}

export interface PlatformResult {
  platform: string;
  status: "success" | "error" | "empty";
  review_count: number;
  reviews: ScrapedReview[];
  error_message: string | null;
  summary_stat?: string | null;
  reply_rate_pct?: number | null;
  /** Real-world places the reviews came from, so the match can be verified. */
  matched_places?: string[];
  /** Places dropped for sitting outside the requested location. */
  excluded_places?: string[];
}

export interface TopicCount {
  topic: string;
  count: number;
}

export interface AnalysisSummary {
  positive_pct: number;
  neutral_pct: number;
  negative_pct: number;
  top_themes: string[];
  top_complaints: TopicCount[];
  summary_text: string | null;
}

export interface CompanyOption {
  id: number;
  name: string;
}

/* ---------- Detailed merged social report ---------- */

export interface ChannelReportSection {
  channel: string;
  summary: string;
  strengths: string[];
  friction_points: string[];
  key_stats: string[];
}

export interface FrictionPoint {
  theme: string;
  source: "scraped" | "consultant" | "both" | string;
  description: string;
  supporting_evidence: string[];
  // Set server-side when the point maps to a counted label.
  related_label?: string | null;
  mention_count?: number | null;
  share_pct?: number | null;
}

/** A cross-channel strength, quantified from the same tables as the friction points.
 *  Older saved reports stored these as bare strings; the backend coerces them on load. */
export interface Strength {
  text: string;
  related_label?: string | null;
  mention_count?: number | null;
  share_pct?: number | null;
}

/** An action tied to the friction point it resolves, so its reach is visible. */
export interface Recommendation {
  text: string;
  /** `theme` of the friction point this addresses, once verified server-side. */
  addresses?: string | null;
  mention_count?: number | null;
  share_pct?: number | null;
}

export interface SentimentBreakdown {
  positive_pct: number;
  neutral_pct: number;
  negative_pct: number;
  classified_count: number;
}

export interface PlatformSentiment {
  platform: string;
  positive_pct: number;
  neutral_pct: number;
  negative_pct: number;
  classified_count: number;
  review_count: number;
  avg_rating?: number | null;
  summary_stat?: string | null;
  reply_rate_pct?: number | null;
}

export interface RatingDistribution {
  one: number;
  two: number;
  three: number;
  four: number;
  five: number;
  total: number;
  average?: number | null;
}

export interface ThemeFrequency {
  label: string;
  count: number;
  share_pct: number;
  negative_pct: number;
  /** Which channels the mentions came from, e.g. { "Google Maps": 6, "Trustpilot": 2 }. */
  by_platform?: Record<string, number>;
}

export interface ThemeVerbatim {
  label: string;
  platform: string;
  text: string;
  rating?: number | null;
  date?: string | null;
  sentiment?: string | null;
  /** Link to the original review, so a quote can be opened at source. */
  url?: string | null;
}

/** Where the star rating and the written sentiment disagree — a rating-inflation signal. */
export interface RatingSentimentCheck {
  both_count: number;
  high_rating_negative: number;
  low_rating_positive: number;
  mismatch_count: number;
  mismatch_pct: number;
}

/** One period on the time axis. `negative_pct` / `avg_rating` are null when that period's
 *  sample is too small to state them — a break in the line, not an invented point. */
export interface TrendPoint {
  period: string;
  label: string;
  reviews: number;
  classified_count: number;
  negative_pct?: number | null;
  avg_rating?: number | null;
}

/** Present only when the collected reviews can support a time axis at all. */
export interface TrendSeries {
  granularity: "month" | "quarter" | string;
  points: TrendPoint[];
  has_sentiment: boolean;
  has_rating: boolean;
  min_sample: number;
}

/** One complaint category's change in share between the earlier and recent windows.
 *  Shares are of all classified feedback in each window, so a rise means the category grew
 *  as a proportion of what customers talked about. */
export interface TopicMovement {
  label: string;
  earlier_count: number;
  recent_count: number;
  earlier_share: number;
  recent_share: number;
  delta: number;
  direction: "rising" | "falling" | "stable" | string;
}

/** Present only when each half of the history carries a real sample. */
export interface TopicMomentum {
  earlier_label: string;
  recent_label: string;
  earlier_total: number;
  recent_total: number;
  items: TopicMovement[];
}

/** Computed server-side in Python — never generated by the LLM. */
export interface ReportEvidence {
  reviews_collected: number;
  /** Reviews carrying a written comment; the rest are star-only ratings. */
  reviews_with_text: number;
  reviews_classified: number;
  overall_sentiment: SentimentBreakdown | null;
  platform_sentiment: PlatformSentiment[];
  rating_distribution: RatingDistribution | null;
  top_themes: ThemeFrequency[];
  top_complaints: ThemeFrequency[];
  verbatims: ThemeVerbatim[];
  /** Null when review volume per period is too thin to plot honestly. */
  trend?: TrendSeries | null;
  /** Which complaint categories are gaining or losing share; null when the history is thin. */
  momentum?: TopicMomentum | null;
  /** Reviews-weighted share of reviews the brand publicly replied to; null if unmeasured. */
  reply_rate_pct?: number | null;
  /** How many reviews that reply rate rests on. */
  reply_rate_base?: number;
  /** Rating-vs-comment disagreement; null when too few reviews carry both. */
  rating_sentiment?: RatingSentimentCheck | null;
}

export interface DetailedSocialReport {
  executive_summary: string;
  overall_sentiment: SentimentBreakdown | null;
  evidence?: ReportEvidence | null;
  channel_breakdown: ChannelReportSection[];
  top_strengths: Strength[];
  top_friction_points: FrictionPoint[];
  website_assessment: string | null;
  recommendations: Recommendation[];
  methodology_note: string;
}

export interface ScrapingResponse {
  brand_name: string;
  scraped_at: string;
  platforms: PlatformResult[];
  total_reviews: number;
  analysis?: AnalysisSummary | null;
  detailed_report?: DetailedSocialReport | null;
  /** The consultant's workbook, echoed back so a past report can be reopened and re-run. */
  manual_analysis?: ManualAnalysisWorkbook | null;
}

export interface PastReportItem {
  filename: string;
  brand_name: string;
  scraped_at: string;
  total_reviews: number;
}

export function sourceLabel(source: string): string {
  if (source === "scraped") return "Scraped reviews";
  if (source === "consultant") return "Manual analysis";
  if (source === "both") return "Both sources";
  return source;
}
